### DISCUSSION TITLE
How does automatic pane naming work? Is it possible to adjust it?

### DISCUSSION THREAD

**Discussion by StigZord** [zellij-org/zellij#2848] [Q&A] [answered]
I noticed that depending on what shell I'm using the way of naming `Pane` is different.

`zsh`:
- When I create new pane its always `Pane X` where X is the pane count

`fish`:
- The pane name changes depend on the path from home, but displays only first letter of each nested folder and full name of last folder
- In VIM though it always displays vim command that opened it and path it was opened in, so for example if I do `vim .` the pane name will be `vim . ~/path/to/where/it/opened` and won't change again

`nushell`:
- When I open zellij first pane will always be named `Pane 1` and all new panes will have name `nu`, no matter if vim is opened or folder changed etc

So my question is, where does Zellij gets the pane name information from? I guess it must be from the shell somehow, if yes can I adjust my shell config to change the way of naming panes when switching folders, or should I create Zellij plugin to handle this?

In case of `vim` I think I should create a small plugin for vim/nvim to change the pane names depending on buffer opened instead, right?

**Comment by danvolchek** [comment, marked as answer]
It seems to be based on terminal title. That can be set through:
```
echo -n -e "\033]0;TEXT\007"
```

Probably every time a command is run (so it updates on cwd changing, or other things). That can be done like this in your .zshrc for zsh:
```
precmd() {
    echo -n -e "\033]0;TEXT\007"
}
```

There's probably more idiomatic ways to set terminal title, but the above at least works for zsh. Unsure about other shells.

It took me forever to find this out through trying random things - it's undocumented. I found this post before I figured it out, and my reaction was literally https://xkcd.com/979/.

**Reply by nd-0r** [reply]
This helped me so much! For the longest time I was trying to do it with `zellij action rename-pane` in .zshrc but it wouldn't work because the session isn't available yet when the .zshrc is run. Thanks!

**Reply by adillari** [reply]
great find! did exactly what I needed

**Reply by pierzchalski** [reply]
For future visitors: it looks like if you _ever_ set a pane's name with `zellij action rename-pane`, Zellij will ignore any future terminal title changes done via `echo -n -e "\e]0;TEXT\a"`. This took me way too long to figure out when I was trying to debug why those `echo`s weren't working for me.
